### PR TITLE
[ONNX] Support device().type() string comparison with constant

### LINKED ISSUE #77671: [ONNX] Exporting the operator prim::type to ONNX opset version 16 is not supported.
### 🐛 Describe the bug

Repro from `pytest test/onnx/test_pytorch_onnx_onnxruntime.py -k test_batched_nms`, with torchvision 0.13.

Error:
```
torch.onnx.symbolic_registry.UnsupportedOperatorError: Exporting the operator prim::type to ONNX opset version 16 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub.
```

### Versions

Collecting environment information...
PyTorch version: 1.12.0a0+gita0bf0f5
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.22.2
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.13.0-28-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080
Nvidia driver version: 510.47.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy==0.950
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.0
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==1.12.0a0+git8de6586
[pip3] torchaudio==0.10.2
[pip3] torchdynamo==0.2.0
[pip3] torchmetrics==0.7.3
[pip3] torchvision==0.13.0a0+c67a583

### LINKED ISSUE #81121: ONNX tests fail after torchvision update
### 🐛 Describe the bug

Namely:
- [ ] test_roi_align
- [ ] test_roi_align_aligned
- [ ] test_roi_pool
- [ ] test_batched_nms

See https://github.com/pytorch/pytorch/runs/7246773625?check_suite_focus=true

### Versions

CI

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86216
* #86169
* __->__ #86168

Fixes `batched_nms` export for torch vision.

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -6621,6 +6636,22 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
     )
 
 
+@_onnx_symbolic("prim::type")
+@_beartype.beartype
+def prim_type(g: jit_utils.GraphContext, device_value: _C.Value, *args, **kwargs):
+    device_node_kind = device_value.node().kind()
+    if device_node_kind == "prim::device":
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:6643]:
```suggestion
    if device_value.node().kind() == "prim::device":
```

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -6621,6 +6636,22 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
     )
 
 
+@_onnx_symbolic("prim::type")
+@_beartype.beartype
+def prim_type(g: jit_utils.GraphContext, device_value: _C.Value, *args, **kwargs):
+    device_node_kind = device_value.node().kind()
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:6642]:
```suggestion
```

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -6621,6 +6636,21 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
...
+def prim_type(g: jit_utils.GraphContext, device_value: _C.Value, *args, **kwargs):
+    if device_value.node().kind() == "prim::device":
+        device = symbolic_helper._try_get_device_from_value(device_value.node().input())
+        if device is not None:
+            return g.op("Constant", value_s=str(device))
+
+    return symbolic_helper._unimplemented(
+        "prim::type",
+        "Device type cannot be statically determined.",
+        device,
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:6650]: Should this be device_value?

Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:6650]: good catch!

Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:6650]:
```suggestion
        device_value,
```

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -6621,6 +6636,21 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
     )
 
 
+@_onnx_symbolic("prim::type")
+@_beartype.beartype
+def prim_type(g: jit_utils.GraphContext, device_value: _C.Value, *args, **kwargs):
+    if device_value.node().kind() == "prim::device":
+        device = symbolic_helper._try_get_device_from_value(device_value.node().input())
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:6643]: Good point, for linking purposes, do we happen to have an TODO issue moving the private methods in symbolic_helper?

[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py]: We don't yet. I can create one.

[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py]: https://github.com/pytorch/pytorch/issues/86181

**Review by justinchuby — APPROVED** [review]
LGTM🚀

**Comment by AllenTiTaiWang** [comment]
> Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
> 
> * [[ONNX] Clarify phrasing of skipScriptTest/skipTraceTest decorators #86216](https://github.com/pytorch/pytorch/pull/86216)
> * [[ONNX] Partially re-enable RoiAlign and RoiPool unit tests #86169](https://github.com/pytorch/pytorch/pull/86169)
> * **->** [[ONNX] Support device().type() string comparison with constant #86168](https://github.com/pytorch/pytorch/pull/86168)
> 
> Fixes #86168

Is the Fixes # referring to the wrong issue in description?

**Review by titaiwangms — APPROVED** [review]

**Review by abock — APPROVED** [review]

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -6621,6 +6636,21 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
     )
 
 
+@_onnx_symbolic("prim::type")
+@_beartype.beartype
+def prim_type(g: jit_utils.GraphContext, device_value: _C.Value, *args, **kwargs):
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:6641]: Need all; or: maybe we should just hide it with _

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -6621,6 +6636,21 @@ def prim_constant(g: jit_utils.GraphContext, *inputs, **attrs):
...
+    if device_value.node().kind() == "prim::device":
+        device = jit_utils.get_device_from_value(device_value.node().input())
+        if device is not None:
+            return g.op("Constant", value_s=str(device))
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:6645]: Thoughts on creating a specially formatted string for device types? (Maybe over complicating)

**Pull request closed without merging**